File: web/src/features/content-items/contentItemMapper.ts
import type { CreateContentItemRequest } from "#/common/api/client"
import type { ContentItemFormValues } from "./ContentItemForm"

export const contentItemMapper = {
  fromFormValues: {
    toCreateRequest: (i: ContentItemFormValues): CreateContentItemRequest => ({
      coverImageUrl: i.coverImageUrl || null,
      format: i.format,
      fullTitle: i.fullTitle,
      isBookmarked: i.isBookmarked,
      isFavorited: i.isFavorited,
      isOngoing: i.isOngoing,
      isSecret: i.isSecret,
      locationType: i.locationType || null,
      locationValue: i.locationValue || null,
      nickName: i.nickName || null,
      placeholderColor: null,
      tags: [...i.tags].map(t => t.trim()).filter(t => t.length > 0),
      totalUnits: i.totalUnits || null,
      unitType: i.unitType,
    }),
  },
}

File: api/src/Yorozu.Presentation/Common/Mappers/ContentItemMapper.cs
using Yorozu.Contracts.ContentItems;
using Yorozu.Domain.ContentItems;

namespace Yorozu.Presentation.Common.Mappers;

internal static class ContentItemMapper {
    extension(ContentItem item) {
        internal ContentItemMiniResponse ToMiniResponse() => new() {
            Title = item.Title,
            CoverImageUrl = item.CoverImageUrl,
            Format = (Contracts.ContentItems.ContentItemFormat)item.Format,
            Id = item.Id,
            IsBookmarked = item.IsBookmarked,
            IsFavorited = item.IsFavorited,
            IsOngoing = item.IsOngoing,
            IsSecret = item.IsSecret,
            LocationType = (Contracts.ContentItems.LocationType?)item.Location?.Type,
            LocationValue = item.Location?.Value,
            PlaceholderLetter = item.PlaceholderLetter,
        };

        internal ContentItemResponse ToResponse() => new() {
            Title = item.Title,
            CoverImageUrl = item.CoverImageUrl,
            Format = (Contracts.ContentItems.ContentItemFormat)item.Format,
            CreatedAt = item.CreatedAt,
            FullTitle = item.FullTitle,
            NickName = item.NickName,
            TotalUnits = item.TotalUnits,
            UnitType = item.UnitType,
            UpdatedAt = item.UpdatedAt,
            Id = item.Id,
            IsBookmarked = item.IsBookmarked,
            IsFavorited = item.IsFavorited,
            IsOngoing = item.IsOngoing,
            IsSecret = item.IsSecret,
            LocationType = (Contracts.ContentItems.LocationType?)item.Location?.Type,
            LocationValue = item.Location?.Value,
            PlaceholderLetter = item.PlaceholderLetter,
            Tags = [.. item.Tags],
        };
    }
}
